### ISSUE TITLE
Related models, rollbacks and afterCreate/afterUpdate events

### ISSUE THREAD

**Issue by sszdh** [phalcon/cphalcon#10450]
Related to this forum post (http://forum.phalconphp.com/discussion/6954/related-models-rollbacks-and-aftercreate-events)

Browsing the source I found that (https://github.com/phalcon/cphalcon/blob/master/phalcon/mvc/model.zep#L2776)

It seams the `_postSave()` calls before `rollback`, so if we have any rollback `afterUpdate` or `afterCreate` already calls anyway!

```php
	/**
	 * Executes internal events after save a record
	 */
	protected function _postSave(boolean success, boolean exists) -> boolean
	{
		if success === true {
			if exists {
				this->fireEvent("afterUpdate");
			} else {
				this->fireEvent("afterCreate");
			}
		}

		return success;
	}
```

But `afterSave` calls after rollback success! (https://github.com/phalcon/cphalcon/blob/master/phalcon/mvc/model.zep#L2795)
I think `_postSave` should be call after rollback and just before emitting `afterSave` event.

Bug o feature?

**Comment by sszdh** [comment]
Okey, Sorry for my delay @andresgutierrez 
So here is example to reproduce:

```php
use Phalcon\DI,
    Phalcon\Mvc\Model,
    Phalcon\Db\Adapter\Pdo\MySQL as Connection,
    Phalcon\Mvc\Model\Manager as ModelsManager,
    Phalcon\Logger\Adapter\File as FileAdapter;


$di = new DI();

$connection = new Connection(array(
    "host"     => "localhost",
    "username" => "root",
    "password" => "****",
    "dbname"   => "phalcon_test",
));

$modelsManager = new ModelsManager();
$modelsManager->setDi($di);
$di['db'] = $connection;
$di['modelsManager'] = $modelsManager;
$di['modelsMetadata'] = new Model\MetaData\Memory();
$di['logger'] = new FileAdapter("test.log"); 

class Robots extends Model
{
    public $id;

    public $name;

    public function initialize()
    {
        $this->hasMany("id", "Parts", "robots_id", ['alias' => 'Parts']);
    }

    public function afterCreate()
    {
        // Sending mail here

        $di = Phalcon\DI::getDefault();
        $di->getLogger()->info("afterCreate Calls so Email sent!");

    }

    public function afterSave()
    {
        $di = Phalcon\DI::getDefault();
        $di->getLogger()->info("afterSave calls.");

    }
}

class Parts extends Model
{

    public $id;

    public $name;

    public function initialize()
    {
        $this->belongsTo("parts_id", "Robots", "id", ['alias' => 'Robot']);
    }
}

try {
    $robot = new Robots();
    $robot->name = 'Sample Robots';
    $part = new Parts();
    $part->name = "Sample Robot Part 1";
    // $part->test_field = "blahblah"; // <-- Do not assign this
    $robot->Parts = [$part];
    $robot->save();

} catch (\Exception $exception) {
    echo $exception->getMessage();
}
```

DB Schema: 

```
CREATE TABLE `parts` (
  `id` int(10) unsigned NOT NULL AUTO_INCREMENT,
  `name` varchar(70) NOT NULL,
  `test_field` varchar(45) NOT NULL, <-- This is the not NULL field that causes Rollback! 
  `robots_id` int(10) NOT NULL,
  PRIMARY KEY (`id`)
) ENGINE=InnoDB DEFAULT CHARSET=latin1;
```
```
CREATE TABLE `robots` (
  `id` int(10) unsigned NOT NULL AUTO_INCREMENT,
  `name` varchar(70) DEFAULT NULL,
  `type` varchar(32) DEFAULT NULL,
  `year` int(11) DEFAULT NULL,
  PRIMARY KEY (`id`)
) ENGINE=InnoDB AUTO_INCREMENT=4 DEFAULT CHARSET=latin1;
```

The log file:
```
[Mon, 27 Jul 15 16:10:46 +0430][INFO] afterCreate Calls so Email sent!
```

**Comment by sszdh** [comment]
Ah! Incidentally it closed :(

**Comment by sszdh** [comment]
Any comments ?

**Comment by andresgutierrez** [comment]
Fixed in the 2.0.x branch, thanks
